controllers/session_test.go

```diff
@@ -0,0 +1,222 @@
+	stdout := gbytes.NewBuffer()
+	g.Expect(simr.Run("session.ls", stdout)).To(gomega.Succeed())
+	g.Expect(sessionCount(stdout)).To(gomega.BeNumerically("==", count))
+}
+
+func TestGetSessionWithKeepAlive(t *testing.T) {
+	g := gomega.NewWithT(t)
+	log := klogr.New()
+	ctrllog.SetLogger(log)
+
+	model := simulator.VPX()
+	model.Cluster = 2
+
+	simr, err := helpers.VCSimBuilder().
+		WithModel(model).Build()
+	if err != nil {
+		t.Fatalf("failed to create VC simulator")
+	}
+	defer simr.Destroy()
+
+	params := session.NewParams().
+		WithServer(simr.ServerURL().Host).
+		WithUserInfo(simr.Username(), simr.Password()).
+		WithFeatures(session.Feature{EnableKeepAlive: true}).
+		WithDatacenter("*")
+
+	// Get first Session
+	s, err := session.GetOrCreate(context.Background(), params)
+	g.Expect(err).ToNot(gomega.HaveOccurred())
+	g.Expect(s).ToNot(gomega.BeNil())
+	AssertSessionCountEqualTo(g, simr, 1)
+
+	// Get session key
+	sessionInfo, err := s.SessionManager.UserSession(context.Background())
+	g.Expect(err).ToNot(gomega.HaveOccurred())
+	g.Expect(sessionInfo).ToNot(gomega.BeNil())
+	firstSession := sessionInfo.Key
+
+	// Get the session again
+	// as keep alive is enabled and session is
+	// not expired we must get the same cached session
+	s, err = session.GetOrCreate(context.Background(), params)
+	g.Expect(err).ToNot(gomega.HaveOccurred())
+	g.Expect(s).ToNot(gomega.BeNil())
+	sessionInfo, err = s.SessionManager.UserSession(context.Background())
+	g.Expect(err).ToNot(gomega.HaveOccurred())
+	g.Expect(sessionInfo).ToNot(gomega.BeNil())
+	g.Expect(sessionInfo.Key).To(gomega.BeEquivalentTo(firstSession))
+	AssertSessionCountEqualTo(g, simr, 1)
+
+	// Try to remove vim session
+	g.Expect(s.Logout(context.Background())).To(gomega.Succeed())
+
+	// after logging out old session must be deleted and
+	// we must get a new different session
+	// total session count must remain 1
+	s, err = session.GetOrCreate(context.Background(), params)
+	g.Expect(err).ToNot(gomega.HaveOccurred())
+	g.Expect(s).ToNot(gomega.BeNil())
+	sessionInfo, err = s.SessionManager.UserSession(context.Background())
+	g.Expect(err).ToNot(gomega.HaveOccurred())
+	g.Expect(sessionInfo).ToNot(gomega.BeNil())
+	g.Expect(sessionInfo.Key).ToNot(gomega.BeEquivalentTo(firstSession))
+	AssertSessionCountEqualTo(g, simr, 1)
+}
+
+func TestGetSessionWithKeepAliveTagManagerLogout(t *testing.T) {
+	g := gomega.NewWithT(t)
+	log := klogr.New()
+	ctrllog.SetLogger(log)
+
+	simulator.SessionIdleTimeout = 1 * time.Second
+	model := simulator.VPX()
+	model.Cluster = 2
+
+	simr, err := helpers.VCSimBuilder().
+		WithModel(model).Build()
+	if err != nil {
+		t.Fatalf("failed to create VC simulator")
+	}
+	defer simr.Destroy()
+
+	params := session.NewParams().
+		WithServer(simr.ServerURL().Host).
+		WithUserInfo(simr.Username(), simr.Password()).
+		WithFeatures(session.Feature{EnableKeepAlive: true, KeepAliveDuration: 2 * time.Second}).WithDatacenter("*")
```

Comment:
The value of `KeepAliveDuration` is the time interval in between send() requests for the purpose of keeping the session alive. So this test does not do what we intend it to do. The misunderstanding is because of the name of the flag.

We will probably later introduce a new flag and deprecate the existing flag due to its misleading name.

Reply:
Ack